backend/lib/richard_burton_web/controllers/publication_controller.ex

```diff
@@ -24,4 +24,17 @@ defmodule RichardBurtonWeb.PublicationController do
 
     conn |> put_status(status) |> json(response_body)
   end
+
+  def parse(conn, %{"csv" => %Plug.Upload{path: path}}) do
+    try do
+      publications = Publication.from_csv!(path)
+      conn |> put_status(:ok) |> json(publications)
+    rescue
+      CSV.RowLengthError ->
+        conn |> put_status(:bad_request) |> json(:incorrect_row_length)
+
+      _ ->
+        conn |> put_status(:bad_request) |> json(:invalid_format)
+    end
+  end
```

Comment:
Is there a particular reason for re-raising in `Publication.from_csv!/1` instead of returning something like `{:ok, publications} | {:error, :invalid_format} | {:error, :incorrect_row_length}`?
Since we're already doing try - rescue in the function, we could free the controller from having to understand the CSV library-specific parsing logic/errors. wdyt?

It could look something like this: 

```
def parse(conn, %{"csv" => %Plug.Upload{path: path}}) do
   case Publication.from_csv!(path) do
     {:ok, publications} -> conn |> put_status(:ok) |> json(publications)
     {:error, reasons} -> conn |> put_status(:bad_request) |> json(:reasons)
   end
end
```

_What if `from_csv/1` didn't return an atom? you may ask_
Since the parse function shouldn't have to return encodable errors either, in the future you could inspect them to compose a message, pattern match on the error types, etc. But I'm mainly thinking that the mapping from the `Library.SomethingError` to `:some_atom_error_in_my_app` perhaps belongs to the parsing side.

Reply:
Yeah I hate catching, reraising, and practically everything involved in this flow.

I attempted to do what you suggest, but there's a major difference in `CSV.decode!/2` and `CSV.decode/2` that made me abandon the refactor of `from_csv!/1` to `from_csv/1`:

- CSV.decode! **fails immediately**
- CSV.decode **inlines errors into the stream**

So at the first glance I would have to consume the whole stream before knowing something failed. There's probably a way to solve it though, so I'll overcome my laziness and give it another shot 😂 

Thank's for the review @jutsuu ✨

Reply:
Gotcha!

And I think you can still keep raising and doing try-rescue in that function. You could just change it to return an error instead of re-raising. In this case, you could argue that is cleaner that way, piping with 
```elixir
foo!() 
|> bar!()
|> Enum.map(..) 
```
rather than pattern matching on each step (something like a `with`). 

Haven't tested any of this but something like this could work:

```elixir
 def from_csv!(path) do
    try do
      path
      |> File.stream!()
      |> CSV.decode!(
        separator: ?;,
        headers: [
          :original_authors,
          :year,
          :country,
          :original_title,
          :title,
          :authors,
          :publisher
        ]
      )
      |> Enum.map(&nest/1)
    rescue
      e in CSV.RowLengthError ->
        {:error, :incorrect_row_length}

      _ -> {:error, :invalid_format}
    end
  end
 ```

Reply:
I'll try to interrupt the stream on the first error, and if I can't manage to do so I'll go for this last suggestion 🙂

Reply:
Decided to use the exceptions. `Enum.reduce_while` does the trick, but the error reporting from the CSV library is nasty, as there's error message but not error code to easily match. And I'd need to catch generic exceptions to identify malformed files anyway 😞 

Here's the code anyway:

```elixir
def from_csv(...) do
    (...)
    |> CSV.decode(...)
    |> Enum.reduce_while([], &reduce_csv_stream/2)
    |> case do
      entries when is_list(entries) ->
        Enum.map(entries, &nest/1)

      {:error, reason} ->
        if String.starts_with?(reason, "Row has length") do
          {:error, :invalid_row_length}
        else
          if String.starts_with?(reason, "Escape sequence expected") do
            {:error, :incorrect_escape_sequence}
          end
        end
    end
  end

  defp reduce_csv_stream({:ok, entry}, acc) do
    {:cont, [entry | acc]}
  end

  defp reduce_csv_stream({:error, reason}, _acc) do
    {:halt, {:error, reason}}
  end
```

Reply:
@jutsuu added the changes in PR https://github.com/wyeworks/richardburton/pull/34 to make my life easier 😅